_posts/2017-10-11-cross-site-request-forgery.markdown

```diff
@@ -0,0 +1,217 @@
+
+The pseudo code for this would look something like:
+
+```java
+public void doGet(HttpServletRequest request, HttpServletResponse response) {
+    token = generateRandomId
+    store token on server side
+    embed token into HTML response as a hidden input field
+}
+
+public void doPost(HttpServletRequest request, HttpServletResponse response) {
+    read token from request
+    if token is same as token stored on server side
+      process request and perform operation
+    else
+      display error and stop processing request
+}
+```
+
+The beauty of this solution is that for a successful exploit the attacker has to include the correct CSRF token in the request, however that should not be possible (assuming that the CRSF tokens are well guarded and are generated using a secure pseudo-random number generator and are tied in one way or another to the logged in user).
+
+Given that random tokens provide a pretty solid protection against CSRF attacks, any kind of an attack that can steal CSRF tokens from a victim has increased value. One very interesting way to steal CSRF tokens is to perform [Path Relative StyleSheet Import] attack for example. Just a friendly reminder that security is a never-ending task...
+
+### CSRF via AJAX
+
+First and foremost we have to learn about the [same-origin policy]: this is a built-in security feature of browsers that **restricts how a document or script loaded from one origin can interact with a resource from another origin**, where *origin* represents a scheme/host/port tuple. One aspect of the policy is that a page under *http://evil.com/index.html* for example has no permission to send requests to *http://example.com/update*, because the origins are different (*http://evil.com:80* vs *http://example.com:80*).
+
+What all of this means is that if the attacker directs an end-user to *evil.com*, they will not be able to make AJAX requests on behalf of the user to *example.com* under normal circumstances. Given that the same-origin policy can be found too restrictive at times, there is a separate feature which allows exceptions from it, called [CORS] (Cross Origin Resource Sharing). The CORS standard defines a set of HTTP headers that can be set by applications to allow AJAX requests from different origins.
+
+Let us assume the worst possible case where the application in question sets the following CORS headers on all of its pages:
+
+```html
+Access-Control-Allow-Origin: http://evil.com
```

Comment:
Did you mean `*` here?

Reply:
No, because you can't combine * origin with allow-credentials :)